### PR TITLE
Enable BN adaptation upon compressed model initialization

### PR DESCRIPTION
**Pull request by vanyalzr**
The PR adds a final BatchNorm statistics adaptation step to any compressed model initializer. 

This is roughly based on the results of the [FNNP paper](https://openreview.net/forum?id=rJeUPlrYvr) and with further considerations that BN statistics adaptation acts as a bias & variance correction procedure following model weight perturbations during compression.

Preliminary results for quantization (ImageNet; fixed random seed for all expts; accuracy measured right after initialization):

| Model        | Quantization bitwidths | Quantization mode | Range initializer |  Accuracy@1 | Accuracy@5 |
| ------------- |:-------------:|:-------------:|:-------------:|:-------------:|:-------------:|
| ResNet18      |  a8w4 | asymmetric, per-channel | mean min max, 100 batches | 49.352 | 73.822 | 
| ResNet18 (BN adapted)      |  a8w4 | asymmetric, per-channel | mean min max, 100 batches | 66.866 | 87.476 | 
| MobilenetV2 | a8w4 |  asymmetric, per-channel | mean min max, 100 batches | 14.034 |  29.278 | 
| MobilenetV2 (BN adapted) | a8w4 |  asymmetric, per-channel | mean min max, 100 batches | 65.216 | 86.304 | 
| MobilenetV2 | a8w8 |  symmetric, per-tensor | mean min max, 100 batches | 69.984 | 89.356 | 
| MobilenetV2 (BN adapted) | a8w8 |  symmetric, per-tensor | mean min max, 100 batches | 70.842 | 89.760 | 
| ResNet50      |  a4w4 | asymmetric, per-channel | mean min max, 100 batches | 2.206 | 5.902 | 
| ResNet50 (BN adapted)   |  a4w4 | asymmetric, per-channel | mean min max, 100 batches | 20.428 | 39.690 | 
| ResNet50      |  a8w4 | asymmetric, per-channel | mean min max, 100 batches | 69.440 | 89.208 | 
| ResNet50 (BN adapted)   |  a8w4 | asymmetric, per-channel | mean min max, 100 batches | 74.866 | 92.244 | 

NB: the "per-channel" tag above relates to weight quantization, activations are always quantized per-tensor. "(A)symmetric" relates to both weights and activations.

Pruning/sparsity results (ImageNet; fixed random seed for all expts; accuracy measured right after initialization)

| Model        | Pruning algo info |  Accuracy@1 | Accuracy@5 |
| ------------- |:-------------:|:-------------:|:-------------:|
| ResNet18 | geometric median criterion, pruning target = 30% | 0.308 | 1.110 | 
| ResNet18 (BN adapted) | geometric median criterion, pruning target = 30% | 33.582 | 59.336 | 
| ResNet18 | magnitude sparsity, sparsity tagret = 70% | 41.196 | 66.952 | 
| ResNet18 (BN adapted) | magnitude sparsity, sparsity target = 70% | 61.222 | 84.022 | 

Preliminary results obtained by fully resetting BN statistics after initialization and running inference of 200 batches (256 samples each).

To do's:
- [x] Add BN adaptation to sparsity & pruning initialization
- [x] Use large momentum value instead of zeroing BN stats completely
- [x] Do iterative adaptation BN layer after BN layer - results in #46

### CODE REVIEW

**Comment by vanyalzr** [comment]
@mkaglins have a look. This initialization procedure can be used for model subjected to filter pruning. We should try to reproduce the results of the FNNP paper which claim that accuracy after BN adaptation correlates well with fine-tuned model accuracy (basically using Figure 4 of the paper as reference).

File: nncf/config_schema.py [hunk 1]
```diff
@@ -402,7 +434,7 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
         },
         # TODO: properly search for the initialize in the compression algo list and
         # remove the necessity to piggyback on the RB sparsity algo entry in the config
-        "initializer": INITIALIZER_SCHEMA,
+        "initializer": QUANTIZATION_INITIALIZER_SCHEMA,
```
Comment (vanyalzr) [inline nncf/config_schema.py:437]: @vshampor is this line still necessary? seems like the initializer schema for RB sparsity was taken from the Quantization one, but there is no initializer schema defined for e.g. Magnitude Sparsity or Filter Pruning. Is the comment above up-to-date?

Comment (vshampor) [reply nncf/config_schema.py:437]: I think it is no longer relevant - we do the initialization of the quantization algo directly during quantization controller creation. The comment may be removed.

**Comment by vanyalzr** [comment]
@mkaglins could you also please run your fine-tuning experiments with L2 and Geomean filter selection criteria separately? It'd be good to know whether the correlation between metrics depends somehow on the pruning criterion chosen.

File: nncf/config_schema.py (hunk 1 above)
Comment (vanyalzr) [reply nncf/config_schema.py:437]: Thanks, got it.

**Comment by vanyalzr** [comment]
Jenkins please retry a build

**Comment by vanyalzr** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

File: nncf/quantization/algo.py [hunk 2]
```diff
@@ -823,6 +826,25 @@ def init_range(self):
                 self._do_range_init(data_loader, num_init_steps, global_init_range_config,
                                     range_init_args.device)
 
+    def run_batchnorm_adaptation(self):
+        init_config = self.quantization_config.get('initializer', {})
+        init_bn_adapt_config = init_config.get('batchnorm_adaptation', {})
+        num_bn_adaptation_steps = init_bn_adapt_config.get('num_bn_adaptation_steps', 200)
+        if num_bn_adaptation_steps < 0:
```
Comment (ljaljushkin) [inline nncf/quantization/algo.py:833]: probably this checks should be the same for filter pruning

File: nncf/initialization.py [hunk 3]
```diff
@@ -106,38 +122,102 @@ def write(cls, msg):
...
+class DataLoaderRangeInitializeRunner(DataLoaderBaseRunner):
+    def __init__(
+            self,
+            model,
+            modules_to_init_vs_init_configs: Dict[str, Tuple[torch.nn.Module, Dict]],
+            init_device: str,
+    ):
+        super().__init__(model, init_device)
+        self.modules_to_init = modules_to_init_vs_init_configs
+        self.progressbar_description = 'Quantizer range initialization'
```
Comment (ljaljushkin) [inline nncf/initialization.py:152]: Quantization specifics shouldn't appear in a base class

File: nncf/config_schema.py [hunk 4]
```diff
@@ -121,9 +121,40 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
...
+}
+
+QUANTIZATION_INITIALIZER_SCHEMA = {
+    "type": "object",
+    "properties": {
+        "batchnorm_adaptation":
+            {
+                "type": "object",
+                "properties": {
+                    "num_bn_adaptation_steps": with_attributes(_NUMBER,
```
Comment (ljaljushkin) [inline nncf/config_schema.py:150]: it makes sense to mention `bn_forget_steps`

File: nncf/initialization.py [hunk 5]
```diff
@@ -106,38 +122,102 @@ def write(cls, msg):
...
+            initializer.apply_init()
+
+
+class DataLoaderBNAdaptationRunner(DataLoaderBaseRunner):
+    def __init__(self, model, init_device: str):
+        super().__init__(model, init_device)
+        self.progressbar_description = 'BatchNorm statistics adaptation'
+        self.num_bn_forget_steps = 5
+        self.momentum_bn_forget = 0.9
+        self.momentum_base = 0.1
```
Comment (ljaljushkin) [inline nncf/initialization.py:179]: looks like this params should come from config

File: nncf/sparsity/magnitude/algo.py [hunk 6]
```diff
@@ -63,6 +67,15 @@ def set_sparsity_level(self, sparsity_level):
 
         self.threshold = self._select_threshold()
         self._set_masks_for_threshold(self.threshold)
+        self.run_batchnorm_adaptation()
+
+    def run_batchnorm_adaptation(self):
+        initializer_params = self.config.get("initializer", {})
```
Comment (ljaljushkin) [inline nncf/sparsity/magnitude/algo.py:73]: why not to move it to BaseAlgoController class, if this initializer is common for all algorithms?

File: nncf/sparsity/magnitude/algo.py [hunk 7]
```diff
@@ -17,11 +17,13 @@
...
+from nncf.initialization import DataLoaderBNAdaptationRunner
 from nncf.nncf_network import NNCFNetwork
 from nncf.sparsity.base_algo import BaseSparsityAlgoBuilder, BaseSparsityAlgoController, SparseModuleInfo
 from nncf.sparsity.layers import BinaryMask
 from nncf.sparsity.magnitude.functions import WEIGHT_IMPORTANCE_FUNCTIONS, calc_magnitude_binary_mask
 from nncf.sparsity.schedulers import SPARSITY_SCHEDULERS
+from nncf.structures import BNAdaptationInitArgs
 
 
 @COMPRESSION_ALGORITHMS.register('magnitude_sparsity')
```
Comment (ljaljushkin) [inline nncf/sparsity/magnitude/algo.py:29]: Is BN adaptation harmful for RB-sparsity? Why did you add it for magnitude one only?

Comment (vanyalzr) [reply nncf/sparsity/magnitude/algo.py]: The sparsification of layers in RB sparsity is not manual like in the magnitude method. For the magnitude method, the adaptation is done once each time a binary mask is constructed. There's no specific point in the training pipeline for RB sparsity where it makes sense to do BN adaptation.

File: nncf/sparsity/magnitude/algo.py (hunk 6 above)
Comment (vanyalzr) [reply nncf/sparsity/magnitude/algo.py]: Agreed. Moved to the base `CompressionAlgorithmController`

File: nncf/config_schema.py (hunk 4 above)
Comment (vanyalzr) [reply nncf/config_schema.py:155]: Done.

File: nncf/quantization/algo.py (hunk 2 above)
Comment (vanyalzr) [reply nncf/quantization/algo.py]: Now done for all algos in the base class

File: nncf/initialization.py (hunk 3 above)
Comment (vanyalzr) [reply nncf/initialization.py]: Renamed appropriately.

**Comment by vanyalzr** [comment]
@AlexKoff88 @vshampor PR is ready to be merged. BN adaptation is switched off by default for now.
